src/rrdp-iostat/rrdp_iostat.ml

```diff
@@ -320,8 +318,78 @@ module Blktap3_stats_wrapper = struct
       low_mem_mode		= false;
     }
 
+  let shm_devices_dir = "/dev/shm"
+
+  let m = Mutex.create ()
+  let cache : ((int * int), int option) Hashtbl.t = Hashtbl.create 100
+
+  let get_pids () =
+    let pid_from_xs (domid, devid) =
+      try
+        let result = with_xs (fun xs ->
+            let path = Printf.sprintf "/local/domain/0/backend/vbd3/%d/%d/kthread-pid" domid devid in
+            Some (int_of_string (xs.Xs.read path))) in
+        result
+      with _ ->
+        None (* Most likely the pid hasn't been written yet *)
+    in
+
+    Mutex.execute m (fun () ->
+        Hashtbl.fold (fun (domid, devid) pid_opt acc ->
+            match pid_opt with
+            | Some pid -> ((domid, devid), pid)::acc
+            | None ->
+              match pid_from_xs (domid, devid) with
+              | Some pid ->
+                Hashtbl.replace cache (domid, devid) (Some pid);
+                ((domid, devid), pid)::acc
+              | None -> acc) cache [])
```

Comment:
Looks a bit confusing having the hashtbl in the same line as the pattern match, maybe put it on its own line so the indentation makes it clearer?